components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/auth/OIDCOauthManager.java

```diff
@@ -62,17 +63,27 @@
     private char[] clientSecret;
 
     public String getClientId() {
-        return clientId;
+        if (clientId != null) {
+            return clientId;
+        } else {
+            ServiceHolder.setOidcOauthManager(new OIDCOauthManager());
+            return ServiceHolder.getOidcOauthManager().getClientId();
+        }
```

Comment:
Since this code can be accessed by multiple threads, we need to synchronize the code for getting a new client id and setting it to the ServiceHolder.

Reply:
Fixed with synchronization.

Reply:
You need to synchronize the whole logic. Otherwise there is no point in synchronizing it this way. Context switch can happen after the if condition is evaluated causing the performDCR() to be called more than once.

And also please simplify the if condition.

Reply:
Resolved as per the comment.